### PR TITLE
Adding time delta argument for Pointcloud generation

### PR DESCRIPTION
**Pull request by jeremyhyde-viam**
Decrease time delta and allowing to be set by user in command line.
Adding in pause before data collection to prevent edge case point cloud generation failure from blocking.

### CODE REVIEW

File: rplidar.go [hunk 1]
```diff
@@ -335,7 +335,7 @@ func (d *Device) NextPointCloud(ctx context.Context) (pointcloud.PointCloud, err
 	defer d.mu.Unlock()
 	pc, timeStamp, err := d.getPointCloud(ctx)
 	if err != nil {
-		return nil, err
+		return nil, nil//err
```
Comment (kkufieta) [inline rplidar.go:338]: why did you change this?

Comment (jeremyhyde-viam) [reply rplidar.go:338]: Since we'll be pinging pointcloud a lot and frequently there is a chance that the rplidar will error out a particular call. This allows us to ignore that missed point cloud and continue without saving it.

Comment (kkufieta) [reply rplidar.go:338]: I see. If that's the desired behavior, then I'd say you should do this:
1) Make that happen in the file that calls this function, e.g. in a file like `cmd/savepcdfiles/main.go`
2) Don't do this preemptively, though. I haven't had any issues with this yet, so I don't see a reason to incorporate this (yet). If it happens to become an issue, though, then we'll have to find a smarter solution than simply ignore an error, e.g. at the very least print out a warning.

Comment (kkufieta) [reply rplidar.go:338]: Revert this back to the way it was.

File: cmd/savepcdfiles/main.go [hunk 2]
```diff
@@ -93,14 +93,19 @@ func savePCDFiles(ctx context.Context, port int, lidarComponent config.Component
...
 		if err != nil {
 			return multierr.Combine(err, myRobot.Close(ctx))
 		}
-		logger.Infow("scanned", "pointcloud_size", pc.Size())
+
+		if pc != nil {
+			logger.Infow("scanned", "pointcloud_size", pc.Size())
+		} else {
+			logger.Infow("warning error reading pcd, skipping")
+		}
```
Comment (kkufieta) [inline cmd/savepcdfiles/main.go:109]: Related to the discussion below, does the else clause get executed at all, given that we're already checking for an error above that?

File: rplidar.go (hunk 1 above)
Comment (jeremyhyde-viam) [reply rplidar.go:338]: Yes, when starting up it would error out due to rplidar c++ stuff. And I believe I also saw this happen several times after ~1 in 20 polls. and that was at 200 ms intervals

File: cmd/savepcdfiles/main.go [hunk 3]
```diff
@@ -91,9 +102,13 @@ func savePCDFiles(ctx context.Context, port int, lidarComponent config.Component
...
 	}
 
+	if !utils.SelectContextOrWait(ctx, time.Second) {
+		return multierr.Combine(ctx.Err(), myRobot.Close(ctx))
+	}
+
 	// Run loop
 	for {
-		if !utils.SelectContextOrWait(ctx, time.Second) {
+		if !utils.SelectContextOrWait(ctx, time.Duration(timeDelta)*time.Millisecond) {
```
Comment (jeremyhyde-viam) [inline cmd/savepcdfiles/main.go:111]: Updated time delta to allow user to change/specify time to wait between measurments

File: cmd/savepcdfiles/main.go [hunk 4]
```diff
@@ -30,7 +31,8 @@ func main() {
 
 // Arguments for the command.
 type Arguments struct {
-	Port       utils.NetPortFlag `flag:"0"`
+	TimeDelta  int               `flag:"0"`
```
Comment (jeremyhyde-viam) [inline cmd/savepcdfiles/main.go:34]: New arg time delta (see below)

File: rplidar.go (hunk 1 above)
Comment (Otterverse) [reply rplidar.go]: A few thoughts here:

Returning nil with no error is not a good idea, especially for a function that's part of the Camera{} interface itself. Someone writing a consumer for that API is going to check for an error, see it's "all good" when there is none, then have their program crash because it tries to reference a nil pointer.

This is also a pretty high level place to ignore all errors.

Since this function actually does a bunch of things, I'd suggest you find lowest level you can where the actual error is being generated and then filter errors there. That is, actually check that the error being generated is specifically the one that you know is "normal" to ignore, and only ignore that one. You don't want to miss if some other type of error happens.

Secondly, and possibly more useful, is when that happens, have it retry whatever function just failed, so that calls still ultimately succeed.

Ex: I'm working on this industrial camera code right now, and when I pull an image buffer from it, it can be "incomplete" so in that case, I just skip it and and pull the next, and the consumer code at the end never sees a problem. But any other sorts of errors can still bubble up to the consumer API call.

File: .gitignore [hunk 5]
```diff
@@ -41,7 +41,6 @@ coverage.txt
 # act
 .secrets
 
-/data*
```
Comment (kkufieta) [inline .gitignore:44]: [q] Why are you removing this, and creating a folder `data` with a gitignore inside instead?

Comment (jeremyhyde-viam) [reply .gitignore:44]: I can remove this, but then users will have to create the folder every time they install it

File: cmd/savepcdfiles/main.go (hunk 4 above)
Comment (kkufieta) [reply cmd/savepcdfiles/main.go:34]: [nit] same as above, can you make the time unit clear?

I love that we can specify the delta in the command line now!

File: cmd/savepcdfiles/main.go [hunk 6]
```diff
@@ -69,10 +80,10 @@ func mainWithArgs(ctx context.Context, args []string, logger golog.Logger) error
 		Attributes: config.AttributeMap{"device_path": usbDevices[0].Path},
 	}
 
-	return savePCDFiles(ctx, int(argsParsed.Port), lidarDevice, logger)
+	return savePCDFiles(ctx, int(argsParsed.TimeDelta), int(argsParsed.Port), lidarDevice, logger)
```
Comment (kkufieta) [inline cmd/savepcdfiles/main.go:83]: [q] Do you need to cast the `argsParsed.TimeDelta` to `int` here? I would assume it's already an integer based on it definition?

File: cmd/savepcdfiles/main.go (hunk 3 above)
Comment (kkufieta) [reply cmd/savepcdfiles/main.go:111]: This is great!

File: cmd/savepcdfiles/main.go [hunk 7]
```diff
@@ -91,9 +102,13 @@ func savePCDFiles(ctx context.Context, port int, lidarComponent config.Component
 		return errors.New("no rplidar found with name: " + name)
 	}
 
+	if !utils.SelectContextOrWait(ctx, time.Second) {
+		return multierr.Combine(ctx.Err(), myRobot.Close(ctx))
+	}
+
```
[the remark this answers is not in the document]
Comment (jeremyhyde-viam) [reply cmd/savepcdfiles/main.go:124]: I figured out that the error with the device came from starting up the point cloud generation too fast. This just adds a 1 sec delay before starting the loop. I also tried adding it in the Start() function in rplidar.go but that didn't work. This change ehre resolves all the errors I've seen thus for with getting point clouds.

File: cmd/savepcdfiles/main.go (hunk 6 above)
Comment (jeremyhyde-viam) [reply cmd/savepcdfiles/main.go]: resolved! I was copying what was done in Port, but I realized thats because it was cast to a special type

File: cmd/savepcdfiles/main.go (hunk 4 above)
Comment (jeremyhyde-viam) [reply cmd/savepcdfiles/main.go:36]: Changed to Milliseconds (Millis has been used before to specify millimeters)

File: cmd/savepcdfiles/main.go [hunk 8]
```diff
@@ -19,9 +19,10 @@ import (
 )
 
 var (
-	defaultPort = 8081
-	logger      = rlog.Logger.Named("save_pcd_files")
-	name        = "rplidar"
+	defaultTimeDelta = 10
```
[the remark this answers is not in the document]
Comment (jeremyhyde-viam) [reply cmd/savepcdfiles/main.go]: done (see below)

File: .gitignore (hunk 5 above)
Comment (jeremyhyde-viam) [reply .gitignore]: I added in a method for generating the folder/clearing it upon start up of a new session. Feel free to compare to your method and choose the best.

File: cmd/savepcdfiles/main.go (hunk 4 above)
Comment (kkufieta) [reply cmd/savepcdfiles/main.go]: Thank you :)

**Review by kkufieta — APPROVED** [review]
LGTM!

**Review by Otterverse — CHANGES_REQUESTED** [review]
One note about using temporary directories, which I think is worth doing properly at this point, but otherwise looks fine to me.

File: cmd/savepcdfiles/main.go [hunk 9]
```diff
@@ -69,10 +82,23 @@ func mainWithArgs(ctx context.Context, args []string, logger golog.Logger) error
...
-	return savePCDFiles(ctx, int(argsParsed.Port), lidarDevice, logger)
+	// Create new data directory
+	newpath := filepath.Join(".", "data")
+
+	err := os.RemoveAll(newpath)
+	if err != nil {
+		return errors.New("error deleting data directory")
+	}
+
+	err = os.MkdirAll(newpath, 0777)
```
[the remark this answers is not in the document]
Comment (jeremyhyde-viam) [reply cmd/savepcdfiles/main.go:93]: So the issue is that, with the current setup for SLAM we are using those saved files. The current process is to generate the saved files, once we have a complete run that we want to test on then we will copy those files over to a different directory that cartographer is looking at. 

I'm fine moving to the temp directory but the deleting of files will have to be on hold until we finish testing SLAM and we start building the full pipeline from RPLiDAR -> SLAM. Additionally, I see the /tmp/ directory is not empty at present

image

Are we able to clear those files everytime we startup a new run on rplidar? I dont want pcd files to get confused with one another

[the remark this answers is not in the document]
Comment (kkufieta) [reply cmd/savepcdfiles/main.go:93]: I think there might be some confusion, based on the fact that Jeremy chose to delete the directory with data every time the program restarts.

The program is currently aimed at collecting data; and we want to keep that since we need to use it in the next step, which is running slam, like Jeremy already said.

I am currently working on a second soon-to-be-opened PR, which will not delete the folder of data. The reason being: I don't want to accidentally delete my hard-earned data just because I restart the rplidar.

Instead, in my upcoming PR, I am allowing the user to define the name of the folder where they want to save their data (the default being `data`), and if the directory already exists, then the new data will be added to the saved data.

I don't mind saving the data in a temp directory - although since it is actually data we want to keep - and keep track of - wouldn't we want to have it in the same directory as the code that's running the rplidar?

**Review by Otterverse — APPROVED** [review]
Per Kat's request in her PR, this is fine to merge to me. It's mostly up to you on the slam team if you want this to delete stuff at startup or not.
